intermediate_source/torch_export_tutorial.py

```diff
@@ -0,0 +1,462 @@
+# we can fix the control flow example above using the ``cond`` op, like so:
+
+from functorch.experimental import control_flow
+
+def bad1_fixed(x):
+    def true_fn(x):
+        return torch.sin(x)
+    def false_fn(x):
+        return torch.cos(x)
+    return control_flow.cond(x.sum() > 0, true_fn, false_fn, [x])
+
+exported_bad1_fixed = export(bad1_fixed, (torch.randn(3, 3),))
+print(exported_bad1_fixed(torch.ones(3, 3)))
+print(exported_bad1_fixed(-torch.ones(3, 3)))
+
+######################################################################
+# There are some limitations one should be aware of:
+# - The predicate (i.e. ``x.sum() > 0``) must result in a boolean or a single-element tensor.
+# - The operands (i.e. ``[x]``) must be tensors.
+# - The branch function (i.e. ``true_fn`` and ``false_fn``) signature must match with the
+# operands and they must both return a single tensor with the same metadata (e.g. dtype, shape, etc.)
+# - Branch functions cannot mutate inputs or globals
+# - Branch functions cannot access closure variables, except for ``self`` if the function is
+# defined in the scope of a method.
+
+######################################################################
+# We can also use ``map``, which applies a function across the first dimension
+# of the first tensor argument.
+
+from functorch.experimental.control_flow import map
+
+def map_example(xs):
+    def map_fn(x, const):
+        def true_fn(x):
+            return x + const
+        def false_fn(x):
+            return x - const
+        return control_flow.cond(x.sum() > 0, true_fn, false_fn, [x])
+    return control_flow.map(map_fn, xs, torch.tensor([2.0]))
+
+exported_map_example= export(map_example, (torch.randn(4, 3),))
+inp = torch.cat((torch.ones(2, 3), -torch.ones(2, 3)))
+print(exported_map_example(inp))
+
+######################################################################
+# Constraints
+# -----------
+# Ops can have different specializations for different tensor shapes, so
+# ``ExportedProgram``s uses constraints on tensor shapes in order to ensure
+# correctness with other inputs.
+# If we try to run the first ``ExportedProgram`` example with a tensor
+# with a different shape, we get an error:
+
+try:
+    exported_mod(torch.randn(10, 100), torch.randn(10, 100))
+except Exception:
+    tb.print_exc()
+
+######################################################################
+# By default, ``torch.export`` requires all tensors to have the same shape
+# as the example inputs, but we can modify the ``torch.export`` call to
+# relax some of these constraints. We use ``torch.export.dynamic_dim`` to
+# express shape constraints manually.
+#
+# We can use ``dynamic_dim`` to remove a dimension's constraints, or to
+# manually provide an upper or lower bound. In the example below, our input
+# ``inp1`` has an unconstrained first dimension, but the size of the second
+# dimension must be in the interval (1, 18].
```

Comment:
not exactly -- we use dynamic_dim to specify which dimensions are dynamic, and can provide shape constraints on top of these dynamic_dims.

Just doing `dynamic_dim(inp1, 0)` tells us that dim 0 of inp1 is dynamic, but doing `3 < dynamic_dim(inp1, 1), dynamic_dim(inp1, 1) <= 18` tells us that the dimension is bounded between (3, 18].